File: 프로그래머스/3/12920. 선입 선출 스케줄링/선입 선출 스케줄링.java
import java.util.*;

class Solution {
    public int solution(int n, int[] cores) {
        
        // n이 코어 개수보다 작거나 같으면 바로 n번째 코어에 배정
        if (n <= cores.length) return n;

        // 0time에 cores.length개 작업은 바로 배정되므로 제외
        n -= cores.length;

        long left = 1;
        // cores 배열 안의 최대값을 구해 n을 곱함
        long right = (long) Arrays.stream(cores).max().getAsInt() * n;

        // 이분 탐색
        while (left <= right) {
            long mid = (left + right) / 2;
            long capacity = 0;
            for (int c : cores) {
                capacity += mid / c;
            }

            if (capacity >= n) {
                right = mid - 1;  // 더 왼쪽에도 답이 있을 수 있음
            } else {
                left = mid + 1;   // 아직 부족하니 오른쪽으로 이동
            }
        }

        // n번째 작업이 완료되기 직전 시간까지 완료된 작업 수
        long doneBefore = 0;
        for (int c : cores) {
            doneBefore += right / c;
        }

        long remain = n - doneBefore; // t 시각에 끝난 작업 중 몇 번째가 n번째인가?

        // t 시각에 동시에 끝나는 코어들 순회
        for (int i = 0; i < cores.length; i++) {
            if ((right+1) % cores[i] == 0) {
                remain--;
                if (remain == 0) {
                    return i + 1; // 코어 번호는 1-based
                }
            }
        }

        return -1; // 논리상 도달 불가
        
    }
}

File: 프로그래머스/3/12920. 선입 선출 스케줄링/선입 선출 스케줄링.py
# 처음 작업은 0초에 시작하여 코어에 배정된다.
def solution(n, cores):
    
    # 코어갯수가 작업갯수보다 같거나 크다면 n번째 코어 return
    if len(cores) >= n:
        return n
    
    # 이분 탐색 구간: 0 ~ (가장 긴 작업 시간) * (남은 작업 수)
    lo, hi = 0, max(cores) * (n - len(cores))
    
    def job_count(time):
        # 각 코어마다 작업이 배정된 횟수
        return sum(time//c + 1 for c in cores)
    
    t = 0 # 마지막 작업이 배정되는 가장 빠른 시간
    while lo <= hi:
        mid = (lo+hi) // 2
        
        if job_count(mid) >= n:
            hi = mid - 1
            t = mid
        else:
            lo = mid + 1
            
    # 직전 시각까지 배정된 개수
    prev = sum((t - 1) // c + 1 for c in cores)
    remain = n - prev  # n번째 작업까지 남은 작업의 횟수
    '''
    t시점엔 무조건 n번째 작업이 배정되는건 확실하다(이분탐색의 이유)
    t-1번째 이후 t시점에 몇개의 작업이 배정될지는 모른다. 
    하지만 그 작업들 중에서 n-prev번째가 n번째 작업이 된다.
    '''
    
    # 시각 t에 작업이 막 끝난(=바로 새 작업 시작 가능한) 코어를 인덱스 오름차로 스캔
    for i, c in enumerate(cores):
        if t % c == 0:
            remain -= 1
            if remain == 0:
                return i + 1  # 1-based 코어 번호
    